pennylane/interfaces/batch/jax.py

```diff
@@ -1,299 +1,332 @@
+        _n (int): a positive integer used to track nesting of derivatives, for example
+            if the nth-order derivative is requested.
+        max_diff (int): If ``gradient_fn`` is a gradient transform, this option specifies
+            the maximum order of derivatives to support. Increasing this value allows
+            for higher order derivatives to be extracted, at the cost of additional
+            (classical) computational overhead during the backwards pass.
+        mode (str): Whether the gradients should be computed on the forward
+            pass (``forward``) or the backward pass (``backward``).
+
+    Returns:
+        list[list[float]]: A nested list of tape results. Each element in
+        the returned list corresponds in order to the provided tapes.
+    """
+    # pylint: disable=unused-argument
+    if max_diff > 1:
+        raise ValueError("The JAX interface only supports first order derivatives.")
+
+    _validate_tapes(tapes)
+
+    for tape in tapes:
+        # set the trainable parameters
+        params = tape.get_parameters(trainable_only=False)
+        tape.trainable_params = qml.math.get_trainable_indices(params)
+
+    parameters = tuple(list(t.get_parameters()) for t in tapes)
+
+    if gradient_fn is None:
+        return _execute_with_fwd(
+            parameters,
+            tapes=tapes,
+            device=device,
+            execute_fn=execute_fn,
+            gradient_kwargs=gradient_kwargs,
+            _n=_n,
+        )
+
+    return _execute(
+        parameters,
+        tapes=tapes,
+        device=device,
+        execute_fn=execute_fn,
+        gradient_fn=gradient_fn,
+        gradient_kwargs=gradient_kwargs,
+        _n=_n,
+    )
+
+
+def _validate_tapes(tapes):
+    """Validates that the input tapes are compatible with JAX support.
+
+    Note: the goal of this validation is to filter out cases where ragged
+    outputs for QNodes may arise. Such QNodes involve creating arrays from
+    ragged nested sequences that can not be handled by JAX.
+
+    Raises:
+        ValueError: if tapes that produce ragged outputs were provided
+    """
+    for t in tapes:
+
+        return_types = [o.return_type for o in t.observables]
+        set_of_return_types = set(return_types)
+        probs_or_sample_measure = Sample in return_types or Probability in return_types
+        if probs_or_sample_measure and len(set_of_return_types) > 1:
+            raise ValueError(
+                "Using the JAX interface, sample and probability measurements cannot be mixed with other measurement types."
+            )
+
+        if Probability in return_types:
+            set_len_wires = set(len(o.wires) for o in t.observables)
+            if len(set_len_wires) > 1:
+                raise ValueError(
+                    "Using the JAX interface, multiple probability measurements need to have the same number of wires specified."
+                )
```

Comment:
@antalszava I had a thought while reading this. I was talking to @trevor-vincent, and there is an issue with the QNode gradient setting. You can't currently do

```python
@qml.qnode(dev, diff_method="best")
def circuit():
    return qml.state()
```

if we swap the `best` method to be `adjoint` by default, since adjoint does not support `state`!

So maybe a solution to this: Let's say we had interfaces/gradient methods raise
```python
raise GradientNotSupportedError(...)
```
internally, if they come across a measurement or a operation that they cannot handle.

The QNode could then do
```python
try:
    # apply gradient method
except GradientNotSupportedError as e:
    if diff_method="best":
        # try the second-best method
    else:
        # user explicitly requested this particular
        # diff method, show the user the exception
        raise e
```

Reply:
This thought doesn't apply so much here, since this is an interface and not a gradient method, but thought I better write this down regardless :)

Reply:
Yes, me too :) Tagging @trbromley as well, since this is relevant to a discussion he and I had earlier.